server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java

```diff
@@ -148,32 +148,48 @@ private Pool createBonusPool(Pool masterPool, List<Pool> existingPools) {
-            if (hostLimited || config.getBoolean(ConfigProperties.STANDALONE)) {
-                virtAttributes.put("unmapped_guests_only", "true");
-            }
-            // Make sure the virt pool does not have a virt_limit,
-            // otherwise this will recurse infinitely
-            virtAttributes.put("virt_limit", "0");
+        if (attributes.containsKey("virt_limit")  && virtQuantity != null) {
+            if (!hasBonusPool(existingPools)) {
+                boolean hostLimited = attributes.containsKey("host_limited") &&
+                    attributes.get("host_limited").equals("true");
+                HashMap<String, String> virtAttributes = new HashMap<String, String>();
+                virtAttributes.put("virt_only", "true");
+                virtAttributes.put("pool_derived", "true");
+                virtAttributes.put("physical_only", "false");
+                if (hostLimited || config.getBoolean(ConfigProperties.STANDALONE)) {
+                    virtAttributes.put("unmapped_guests_only", "true");
+                }
+                // Make sure the virt pool does not have a virt_limit,
+                // otherwise this will recurse infinitely
+                virtAttributes.put("virt_limit", "0");
 
-            // Favor derived products if they are available
-            Product sku = masterPool.getDerivedProduct() != null ? masterPool.getDerivedProduct() :
-                masterPool.getProduct();
+                // Favor derived products if they are available
+                Product sku = masterPool.getDerivedProduct() != null ? masterPool.getDerivedProduct() :
+                    masterPool.getProduct();
 
-            // Using derived here because only one derived pool is created for
-            // this subscription
-            Pool bonusPool = PoolHelper.clonePool(masterPool, sku, virtQuantity, virtAttributes, "derived",
-                prodCurator, null);
+                // Using derived here because only one derived pool is created for
+                // this subscription
+                Pool bonusPool = PoolHelper.clonePool(masterPool, sku, virtQuantity, virtAttributes,
+                    "derived", prodCurator, null);
 
-            log.info("Creating new derived pool: {}", bonusPool);
-            return bonusPool;
+                log.info("Creating new derived pool: {}", bonusPool);
+                return bonusPool;
+            }
+            else {
+                Pool bonus = getBonusPool(existingPools);
+                long virtLong = -1L;
+                try {
+                    virtLong = Long.parseLong(virtQuantity);
+                }
+                catch (NumberFormatException pe) {
+                    // it is unlimited, leave it at -1L
+                }
+                if (virtLong != bonus.getQuantity()) {
+                    bonus.setQuantity(virtLong);
+                    poolManager.setPoolQuantity(bonus, virtLong);
+                    return bonus;
```

Comment:
first, since this method now updates pool as well, we should rename it to `createOrUpdateBonusPool` from `createBonusPool`

also, why return the bonus pool if we are updating it? the `poolManager.setPoolQuantity` method merges the change, if that is the concern.

returning the pool here calls `poolManager.createPool` , which:
* logs that the update as "new pool"
* calls poolCurator.create(pool) ( which would probably cause an error )
* emit a pool create event instead of pool update event

I think we should simply not return in the else case and simply emit an update event, thoughts? am I missing something?